src/ImageSharp.Web/Resolvers/PhysicalFileSystemCacheResolver.cs

```diff
@@ -30,21 +30,23 @@ public PhysicalFileSystemCacheResolver(FileInfo metaFileInfo, FormatUtilities fo
         }
 
         /// <inheritdoc/>
-        public async Task<ImageCacheMetadata> GetMetaDataAsync()
-        {
-            using Stream stream = OpenFileStream(this.metaFileInfo.FullName);
-            this.metadata = await ImageCacheMetadata.ReadAsync(stream);
-            return this.metadata;
-        }
+        public async Task<ImageCacheMetadata> GetMetaDataAsync() => this.metadata ??= await this.ReadMetadataAsync();
 
         /// <inheritdoc/>
-        public Task<Stream> OpenReadAsync()
+        public async Task<Stream> OpenReadAsync()
         {
+            ImageCacheMetadata metadata = await this.GetMetaDataAsync();
             string path = Path.ChangeExtension(
                 this.metaFileInfo.FullName,
-                this.formatUtilities.GetExtensionFromContentType(this.metadata.ContentType));
+                this.formatUtilities.GetExtensionFromContentType(metadata.ContentType));
 
-            return Task.FromResult(OpenFileStream(path));
+            return OpenFileStream(path);
+        }
+
+        private async Task<ImageCacheMetadata> ReadMetadataAsync()
+        {
+            using Stream stream = OpenFileStream(this.metaFileInfo.FullName);
+            return await ImageCacheMetadata.ReadAsync(stream);
         }
 
         private static Stream OpenFileStream(string path)
```

Comment:
We should add aggressive inlining here also and make the result return a `FileStream`.